### DISCUSSION TITLE
How to manipulate response body in ASP.Net core Acton filter without affecting other middlewares?

### DISCUSSION THREAD

**Discussion by rbipin** [dotnet/aspnetcore#22154] [General]
I have two endpoints and the result needs to be formatted differently for each endpoint 
I'm manipulating the response body in _OnResultExecuting_ in an Action Filter. 
My API has a couple of middlewares that are also changing response status codes. Since my Action filter is already doing _response.body.writeAsync_. The middlewares are throwing error that the _response has already started_. Is there a work around to safely manipulate the response without affecting any other middlewares. The code snippet is given below.

```csharp
      var response = context.HttpContext.Response;
            if (response.StatusCode != 200)
            {
                base.OnResultExecuting(context);
            }
            var result = context.Result as ObjectResult;
            var serializedContent = result.Value.Serialize();
            var data = Encoding.UTF8.GetBytes(serializedContent);
            response.Headers.Add("Content-Type", "application/xml");
            await response.Body.WriteAsync(data, 0, data.Length);
           base.OnResultExecuting(context);
`

**Comment by davidfowl** [comment]
Only one middleware can write to the response. You have a couple of options:
- Buffer the response. If you buffer the response in a middleware then that middleware can also flush the response to the output. 
- The other middleware should no-op when the response has already started. 

What middleware is in the pipeline?
